File: state/state.go
// Package state implements the Trellis State model (spec §4): the live lifecycle
// condition of a resource, *derived* from its two projections plus health and
// the reconciler's control stance —
//
//	state = f(desired, observed, health)
//
// State is a pure, recomputable function, never stored as ground truth. That is
// "no magic" applied to state: every state is explainable by showing its inputs.
package state

import (
	"time"

	"github.com/hakkiki/trellis/model"
)

// Sync is the observed-vs-desired sub-dimension (spec §4).
type Sync int

const (
	// SyncUnknown means observation is stale or missing — we cannot assert.
	SyncUnknown Sync = iota
	// SyncInSync means observed matches desired.
	SyncInSync
	// SyncProgressing means observed differs from desired because a newer
	// authored generation is still rolling out (good).
	SyncProgressing
	// SyncDrifted means observed diverged from the generation already applied,
	// with nothing newer pending — an unauthored change (bad).
	SyncDrifted
)

// Control is the reconciler's stance toward a resource (spec §4).
type Control int

const (
	// ControlSettled means the reconciler is holding; no action in flight.
	ControlSettled Control = iota
	// ControlReconciling means the reconciler is actively closing a gap.
	ControlReconciling
	// ControlStalled means convergence cannot make progress — needs a human.
	ControlStalled
	// ControlFrozen means reconciliation is suspended (break-glass, §7).
	ControlFrozen
)

// State is the named lifecycle state — the salient cell of the
// Sync × Health × Control product (spec §4).
type State int

const (
	// StateUnknown means observation is stale/missing; the system cannot assert
	// a state and must fail safe (Invariant 7).
	StateUnknown State = iota
	// StateConverged means InSync · Healthy · Settled — the goal.
	StateConverged
	// StateConverging means an authored gap is closing (a deploy).
	StateConverging
	// StateDrifted means an unauthored divergence the reconciler will correct.
	StateDrifted
	// StateDegraded means correct-but-unhealthy → self-heal.
	StateDegraded
	// StateStalled means convergence is blocked — needs a human.
	StateStalled
	// StateFrozen means reconciliation is suspended (break-glass) with a debt.
	StateFrozen
)

func (s State) String() string {
	switch s {
	case StateConverged:
		return "Converged"
	case StateConverging:
		return "Converging"
	case StateDrifted:
		return "Drifted"
	case StateDegraded:
		return "Degraded"
	case StateStalled:
		return "Stalled"
	case StateFrozen:
		return "Frozen"
	default:
		return "Unknown"
	}
}

// Derive computes the named State of a resource from its desired and observed
// projections, the reconciler's control stance, and a freshness check. It is the
// pure function at the heart of the State model (spec §4) and is never stored.
//
// stalenessBudget bounds how old an observation may be before it is treated as
// Unknown rather than assumed-converged (Invariant 7, fail-safe-on-Unknown). A
// zero budget disables the freshness check.
func Derive(d model.DesiredResource, o model.Observation, ctrl Control, now time.Time, stalenessBudget time.Duration) State {
	// A break-glass freeze dominates everything: reconciliation is suspended.
	if ctrl == ControlFrozen {
		return StateFrozen
	}

	sync := classifySync(d, o, now, stalenessBudget)

	// A reconciler that has given up (flap/blast-radius breaker tripped, or
	// repeated failure) reports Stalled — but only once it is actually behind.
	if ctrl == ControlStalled && sync != SyncInSync {
		return StateStalled
	}

	// Fail-safe on Unknown: never assume Converged from stale data.
	if sync == SyncUnknown {
		return StateUnknown
	}

	switch sync {
	case SyncInSync:
		switch o.Health {
		case model.HealthHealthy:
			return StateConverged
		case model.HealthDegraded:
			return StateDegraded
		default:
			// InSync but liveness unknown — cannot assert healthy.
			return StateUnknown
		}
	case SyncProgressing:
		return StateConverging
	case SyncDrifted:
		return StateDrifted
	}
	return StateUnknown
}

// classifySync decides the observed-vs-desired relationship, using provenance
// (generations), not gap size, to separate authored progress from drift (the
// key non-obvious rule of spec §4).
func classifySync(d model.DesiredResource, o model.Observation, now time.Time, stalenessBudget time.Duration) Sync {
	if !o.Exists {
		// Desired exists but nothing is observed yet: an authored creation is in
		// flight.
		return SyncProgressing
	}
	if stalenessBudget > 0 && !o.ObservedAt.IsZero() && now.Sub(o.ObservedAt) > stalenessBudget {
		return SyncUnknown
	}
	if model.SpecEqual(o.Spec, d.Spec) {
		return SyncInSync
	}
	// observed != desired — provenance decides.
	if d.Generation > o.AppliedGeneration {
		// A newer authored generation has not been applied yet: progress.
		return SyncProgressing
	}
	// observed diverged from the generation already applied: unauthored drift.
	return SyncDrifted
}

File: app/src/sim/state.ts
// The Trellis State model (spec §4): state = f(desired, observed, health).
// A pure, recomputable function — never stored as ground truth.

import { type DesiredResource, type Observation, specEqual } from "./model";

export type Sync = "Unknown" | "InSync" | "Progressing" | "Drifted";
export type Control = "Settled" | "Reconciling" | "Stalled" | "Frozen";
export type State =
  | "Unknown"
  | "Converged"
  // Intentionally parked to save cost (docs: Cost & parity) — a good steady
  // state, not down and not drift; the reconciler holds, never wakes it.
  | "Dormant"
  | "Converging"
  | "Drifted"
  | "Degraded"
  | "Stalled"
  | "Frozen"
  // Stateful quorum roll-up (§4): minority of nodes serving
  | "Unavailable"
  // Job lifecycle (terminal progression, §4)
  | "Pending"
  | "Running"
  | "Succeeded"
  | "Failed";

export const ALL_STATES: State[] = [
  "Converged",
  "Dormant",
  "Converging",
  "Drifted",
  "Degraded",
  "Unavailable",
  "Stalled",
  "Frozen",
  "Unknown",
];

/** Tailwind/CSS color token for each state (matches src/styles/global.css). */
export function stateColorVar(s: State): string {
  switch (s) {
    case "Converged":
    case "Succeeded":
      return "var(--state-converged)";
    case "Converging":
    case "Running":
      return "var(--state-converging)";
    case "Dormant":
      // Intentionally suspended — reuse the calm "held" token, not an alarm color.
      return "var(--state-frozen)";
    case "Degraded":
      return "var(--state-degraded)";
    case "Drifted":
      return "var(--state-drifted)";
    case "Stalled":
    case "Failed":
      return "var(--state-stalled)";
    case "Unavailable":
      return "var(--state-unavailable)";
    case "Frozen":
      return "var(--state-frozen)";
    case "Pending":
      return "var(--state-unknown)";
    default:
      return "var(--state-unknown)";
  }
}

/** Severity ordering for rolling a Frame's state up from its children (§4). */
const SEVERITY: Record<State, number> = {
  Converged: 0,
  Succeeded: 0,
  // Parked is a good steady state — benign, but visible in a roll-up so a frame
  // with idle capacity reads "Dormant" rather than masquerading as fully live.
  Dormant: 1,
  Pending: 1,
  Running: 1,
  Unknown: 2,
  Converging: 3,
  Drifted: 4,
  Degraded: 5,
  Frozen: 6,
  Unavailable: 7,
  Stalled: 8,
  Failed: 8,
};

/** Roll a Frame's state up from its children's states — worst-of (§4). */
export function rollup(states: State[]): State {
  if (!states.length) return "Unknown";
  return states.reduce<State>(
    (worst, s) => (SEVERITY[s] > SEVERITY[worst] ? s : worst),
    "Converged",
  );
}

/** A Job's State is its terminal progression; completion is success, not drift. */
function deriveJob(o: Observation): State {
  if (!o.exists || !o.phase) return "Pending";
  switch (o.phase) {
    case "pending":
      return "Pending";
    case "running":
      return "Running";
    case "succeeded":
      return "Succeeded";
    case "failed":
      return "Failed";
  }
}

/** An External workload is observe-only: its State reflects the observed health,
 *  and the reconciler never converges it. */
function deriveExternal(o: Observation): State {
  if (!o.exists) return "Unknown";
  if (o.health === "Degraded") return "Degraded";
  if (o.health === "Healthy") return "Converged";
  return "Unknown";
}

function classifySync(
  d: DesiredResource,
  o: Observation,
  nowMs: number,
  stalenessBudgetMs: number,
): Sync {
  if (!o.exists) return "Progressing"; // authored creation in flight
  if (stalenessBudgetMs > 0 && o.observedAtMs > 0 && nowMs - o.observedAtMs > stalenessBudgetMs) {
    return "Unknown";
  }
  if (specEqual(o.spec, d.spec)) return "InSync";
  // observed != desired — provenance (generation), not gap size, decides (§4).
  if (d.generation > o.appliedGeneration) return "Progressing";
  return "Drifted";
}

/**
 * Derive the named lifecycle State. stalenessBudgetMs bounds how old an
 * observation may be before it is treated as Unknown rather than
 * assumed-converged (Invariant 7). Zero disables the freshness check.
 */
export function derive(
  d: DesiredResource,
  o: Observation,
  control: Control,
  nowMs: number,
  stalenessBudgetMs: number,
): State {
  // Jobs and External workloads have different lifecycles (§1, §4).
  if (d.lifecycle === "job") return deriveJob(o);
  if (d.lifecycle === "external") return deriveExternal(o);

  if (control === "Frozen") return "Frozen";

  // Intentionally parked to save cost (docs: Cost & parity). Checked before the
  // sync/health/quorum classification so a paused DB (quorum 0) or a scaled-to-
  // zero service reads Dormant, never Unavailable/Degraded — and never drift.
  if (o.dormant) return "Dormant";

  // Warming back up after a park (cold start, docs: Cost & parity guardrail 4) —
  // benign progress toward live, not drift. The reconciler holds while it warms.
  if (o.resuming) return "Converging";

  const sync = classifySync(d, o, nowMs, stalenessBudgetMs);

  // A reconciler that has given up (circuit breaker tripped) reports Stalled
  // until the resource is genuinely Converged again (the breaker resets).
  if (control === "Stalled") {
    return sync === "InSync" && o.health === "Healthy" ? "Converged" : "Stalled";
  }
  if (sync === "Unknown") return "Unknown";

  switch (sync) {
    case "InSync":
      // Stateful clusters roll health up by quorum (§4).
      if (d.lifecycle === "stateful") return statefulHealth(o);
      if (o.health === "Healthy") return "Converged";
      if (o.health === "Degraded") return "Degraded";
      return "Unknown";
    case "Progressing":
      return "Converging";
    case "Drifted":
      return "Drifted";
  }
  return "Unknown";
}

/** Quorum roll-up for a stateful cluster: all nodes → Converged; a majority
 *  still serving → Degraded; a minority → Unavailable (quorum lost). */
function statefulHealth(o: Observation): State {
  const q = o.quorum ?? { healthy: o.health === "Healthy" ? 1 : 0, total: 1 };
  if (q.healthy >= q.total) return "Converged";
  const majority = Math.floor(q.total / 2) + 1;
  return q.healthy >= majority ? "Degraded" : "Unavailable";
}
